Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs

```diff
@@ -379,6 +376,31 @@ private Task<BoolResult> RestoreCheckpointIncrementalAsync(OperationContext cont
                 });
         }
 
+        private async Task RestoreFilesAsync(OperationContext context, AbsolutePath checkpointTargetDirectory, Dictionary<string, string> newCheckpointInfo)
+        {
+            if (_configuration.IncrementalCheckpointDegreeOfParallelism <= 1)
+            {
+                foreach (var (key, value) in newCheckpointInfo)
+                {
+                    await RestoreFileAsync(context, checkpointTargetDirectory, key, value).ThrowIfFailure();
+                }
+
+            }
+            else
+            {
+                await ParallelAlgorithms.WhenDoneAsync(
+                    _configuration.IncrementalCheckpointDegreeOfParallelism,
+                    context.Token,
+                    action: async (addItem, kvp) =>
+                    {
+                        var key = kvp.Key;
+                        var value = kvp.Value;
+                        await RestoreFileAsync(context, checkpointTargetDirectory, key, value).ThrowIfFailure();
+                    },
+                    items: newCheckpointInfo.ToArray());

```

Comment:
Question: what happens if one of the parallel restores fails here?

Reply:
In this case WhenDoneAsync will fail with aggregateexception.

---
In reply to: [325331241](https://github.com/microsoft/BuildXL/pull/882#discussion_r325331241) [](ancestors = 325331241)